signer/src/api/info.rs

```diff
@@ -0,0 +1,505 @@
+                signer_tip: None,
+                node_tip: None,
+                node_chain: None,
+                node_version: None,
+                node_subversion: None,
+            },
+            stacks: StacksInfo {
+                signer_tip: None,
+                node_tip: None,
+                node_bitcoin_block_height: None,
+                node_version: None,
+            },
+            dkg: DkgInfo {
+                rounds: 0,
+                current_aggregate_key: None,
+                target_rounds: 1,
+                min_bitcoin_block_height: None,
+            },
+            build_info: BuildInfo {
+                rust_version: crate::RUSTC_VERSION.to_string(),
+                git_revision: crate::GIT_COMMIT.to_string(),
+                target_arch: crate::TARGET_ARCH.to_string(),
+                target_env_abi,
+            },
+            timestamp: time::OffsetDateTime::now_utc().to_string(),
+        }
+    }
+}
+
+impl IntoResponse for InfoResponse {
+    fn into_response(self) -> axum::response::Response {
+        Json(self).into_response()
+    }
+}
+
+/// Handler for the `/info` endpoint. This method is infallible and returns
+/// `null` for any missing information.
+pub async fn info_handler<C: Context>(state: State<ApiState<C>>) -> InfoResponse {
+    let bitcoin_client = state.ctx.get_bitcoin_client();
+    let stacks_client = state.ctx.get_stacks_client();
+    let storage = state.ctx.get_storage();
+    let config = state.ctx.config();
+
+    let mut response = InfoResponse::default();
+
+    response.populate_local_chain_info(&storage).await;
+    response.populate_bitcoin_node_info(&bitcoin_client).await;
+    response.populate_stacks_node_info(&stacks_client).await;
+    response.populate_dkg_info(&storage, config).await;
+
+    response
+}
+
+impl InfoResponse {
+    /// Populates the local Bitcoin and Stacks chain tip information.
+    async fn populate_local_chain_info(&mut self, storage: &impl DbRead) {
+        match storage.get_bitcoin_canonical_chain_tip().await {
+            Ok(Some(local_bitcoin_chain_tip)) => {
+                let bitcoin_block = storage
+                    .get_bitcoin_block(&local_bitcoin_chain_tip)
+                    .await
+                    .inspect_err(|e| {
+                        tracing::error!("error reading bitcoin block from the database: {}", e)
+                    });
+
+                let Ok(Some(bitcoin_block)) = bitcoin_block else {
+                    tracing::error!(
+                        "canonical tip found but could not retrieve block from the database"
+                    );
+                    return;
+                };
+
+                self.bitcoin.signer_tip = Some(ChainTipInfo {
+                    block_hash: bitcoin_block.block_hash,
+                    block_height: bitcoin_block.block_height,
+                });
+
+                match storage
+                    .get_stacks_chain_tip(&bitcoin_block.block_hash)
```

Comment:
This is (currently) a very expensive query (no context window constraint), are we okay with having it triggerable so easily?

Reply:
Was this one of them? I can't remember -- this one has the `LIMIT 1` so I thought it wouldn't be a problem.

Reply:
We have the limit outside, but no limits on the CTE. Checked the execution time of `get_stacks_chain_tip` and it seems the limit it's not helping much:
```
---- postgres_bech::bench_get_stacks_tip::playnet stdout ----
Executed query in 77.858541ms

---- postgres_bech::bench_get_stacks_tip::mainnet stdout ----
Executed query in 123.353708ms

---- postgres_bech::bench_get_stacks_tip::synth_small stdout ---- // bitcoin_blocks: 50, stacks_blocks_per_anchor: 20
Executed query in 1.705084ms

---- postgres_bech::bench_get_stacks_tip::synth stdout ---- // bitcoin_blocks: 1000, stacks_blocks_per_anchor: 20
Executed query in 61.17175ms
```

Reply:
Those look "OK" anyway, at least for now.

The query made here shouldn't lock/block anything, and there's nothing waiting on it other than the caller, so I think this should be okay for the time being. My hope is that we've managed to refactor all of this db stuff in the semi-near future...